pkg/ingester/owned_series.go

```diff
@@ -209,3 +201,100 @@ func secondaryTSDBHashFunctionForUser(userID string) func(labels.Labels) uint32
 		return mimirpb.ShardByAllLabels(userID, ls)
 	}
 }
+
+type ownedSeriesIngesterRingStrategy struct {
+	instanceID    string
+	ingestersRing ring.ReadRing
+
+	getIngesterShardSize func(user string) int
+
+	previousRing ring.ReplicationSet
+}
+
+func newOwnedSeriesIngesterRingStrategy(instanceID string, ingesterRing ring.ReadRing, getIngesterShardSize func(user string) int) *ownedSeriesIngesterRingStrategy {
+	return &ownedSeriesIngesterRingStrategy{
+		instanceID:           instanceID,
+		ingestersRing:        ingesterRing,
+		getIngesterShardSize: getIngesterShardSize,
+	}
+}
+
+func (ir *ownedSeriesIngesterRingStrategy) ownerKeyAndValue() (string, string) {
+	return "ingester", ir.instanceID
+}
+
+func (ir *ownedSeriesIngesterRingStrategy) checkRingForChanges() (bool, error) {
+	rs, err := ir.ingestersRing.GetAllHealthy(ownedSeriesRingOp)
+	if err != nil {
+		return false, err
+	}
+
+	// Ignore state and IP address changes, since they have no impact on token distribution
+	ringChanged := ring.HasReplicationSetChangedWithoutStateOrAddr(ir.previousRing, rs)
+	ir.previousRing = rs
+	return ringChanged, nil
+}
+
+func (ir *ownedSeriesIngesterRingStrategy) shardSizeForUser(userID string) int {
+	return ir.getIngesterShardSize(userID)
+}
+
+func (ir *ownedSeriesIngesterRingStrategy) tokenRangesForUser(userID string, shardSize int) (ring.TokenRanges, error) {
+	subring := ir.ingestersRing.ShuffleShard(userID, shardSize)
+
+	ranges, err := subring.GetTokenRangesForInstance(ir.instanceID)
+	if errors.Is(err, ring.ErrInstanceNotFound) {
+		// Not an error.
+		return nil, nil
+	}
+
+	return ranges, err
+}
+
+type ownedSeriesPartitionRingStrategy struct {
+	partitionID          int32
+	partitionRingWatcher *ring.PartitionRingWatcher
+
+	getPartitionShardSize func(user string) int
+
+	previousActivePartition []int32
+}
+
+func newOwnedSeriesPartitionRingStrategy(partitionID int32, partitionRing *ring.PartitionRingWatcher, getPartitionShardSize func(user string) int) *ownedSeriesPartitionRingStrategy {
+	return &ownedSeriesPartitionRingStrategy{
+		partitionID:           partitionID,
+		partitionRingWatcher:  partitionRing,
+		getPartitionShardSize: getPartitionShardSize,
+	}
+}
+
+func (pr *ownedSeriesPartitionRingStrategy) checkRingForChanges() (bool, error) {
+	// When using partitions ring, we consider ring to be changed if active partitions have changed.
+	r := pr.partitionRingWatcher.PartitionRing()
+	activePartitions := r.ActivePartitionIDs()
```

Comment:
[nit, not blocker] **IIRC**, so far we didn't need `ActivePartitionIDs()` to return a sorted list. It was sorting ids just to simplify tests. Can you add a comment in dskit to mention it returns sorted list of ids and eventually a unit test to enforce it (don't remember if we already have it).

Reply:
Done in https://github.com/grafana/dskit/pull/503